src/MagnumPlugins/BasisImporter/BasisImporter.cpp

```diff
@@ -0,0 +1,217 @@
+#include <Corrade/Containers/Optional.h>
+#include <Corrade/Utility/Debug.h>
+#include <Magnum/PixelFormat.h>
+#include <Magnum/Trade/ImageData.h>
+
+namespace Magnum { namespace Trade {
+
+namespace {
+
+/* Map BasisTranscodingType to Magnum::CompressedPixelFormat */
+CompressedPixelFormat getTextureFormat(BasisTranscodingType type, bool hasAlpha) {
+    switch(type) {
+        case BasisTranscodingType::Etc1:
+        case BasisTranscodingType::Etc2:
+            return (hasAlpha) ? CompressedPixelFormat::Etc2RGBA8Unorm
+                : CompressedPixelFormat::Etc2RGB8Unorm;
+        case BasisTranscodingType::Bc1:
+            return (hasAlpha) ? CompressedPixelFormat::Bc1RGBAUnorm
+                : CompressedPixelFormat::Bc1RGBUnorm;
+        case BasisTranscodingType::Bc3:
+            CORRADE_ASSERT(hasAlpha, "Bc3 always has alpha, but non-alpha was requested",
+                CompressedPixelFormat::Bc3RGBAUnorm);
+            return CompressedPixelFormat::Bc3RGBAUnorm;
+        case BasisTranscodingType::Bc4:
+            return CompressedPixelFormat::Bc4RUnorm;
+        case BasisTranscodingType::Bc5:
+            return CompressedPixelFormat::Bc5RGUnorm;
+        case BasisTranscodingType::Bc7M6OpaqueOnly:
+            CORRADE_ASSERT(hasAlpha, "Bc7 is opaque only, but alpha was requested",
+                CompressedPixelFormat::Bc7RGBAUnorm);
+            return CompressedPixelFormat::Bc7RGBAUnorm;
+        case BasisTranscodingType::Pvrtc1_4OpaqueOnly:
+            CORRADE_ASSERT(hasAlpha, "Pvrtc1 is opaque only, but alpha was requested",
+                CompressedPixelFormat::PvrtcRGB4bppUnorm);
+            return CompressedPixelFormat::PvrtcRGB4bppUnorm;
+        default:
+            CORRADE_ASSERT_UNREACHABLE();
+    }
+}
+
+basist::transcoder_texture_format getTranscoderTextureFormat(BasisTranscodingType type) {
+    constexpr basist::transcoder_texture_format formats[] = {
+        basist::transcoder_texture_format::cTFETC1,
+        basist::transcoder_texture_format::cTFETC2,
+        basist::transcoder_texture_format::cTFBC1,
+        basist::transcoder_texture_format::cTFBC3,
+        basist::transcoder_texture_format::cTFBC4,
+        basist::transcoder_texture_format::cTFBC5,
+        basist::transcoder_texture_format::cTFBC7_M6_OPAQUE_ONLY,
+        basist::transcoder_texture_format::cTFPVRTC1_4_OPAQUE_ONLY
+    };
+
+    return formats[Int(type)];
+}
+
+}
+
+struct BasisImporter::State {
+    /* There is only this type of codebook */
+    basist::etc1_global_selector_codebook _codebook;
+    basist::basisu_transcoder _transcoder{&_codebook};
+
+    State(): _codebook(basist::g_global_selector_cb_size,
+        basist::g_global_selector_cb) {}
+};
+
+BasisImporter::BasisImporter() = default;
+
+BasisImporter::BasisImporter(PluginManager::AbstractManager& manager, const std::string& plugin): AbstractImporter{manager, plugin} {
+    /* Does nothing if already initialized */
+    basist::basisu_transcoder_init();
```

Comment:
~~Maybe move this line to the State constructor? Since that one does similar initialization stuff.~~

Actually, since this can be done just once, what about putting it into a static `initialize()` function instead? (See `FreeTypeFont` for an example.) Is there a corresponding `deinit()` function?

Reply:
There is no deinit function and successive calls to init() [don't do anything](https://github.com/BinomialLLC/basis_universal/blob/master/transcoder/basisu_transcoder.cpp#L1660)

Reply:
Okay, thank you -- can you then move it to the State constructor to be together with similar setup calls?

*(Oh god, a `static` function-local variable :see_no_evil: -- I need to double check the supposed thread safety of this thing.)*

Reply:
Checking https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables, there doesn't seem to be any special case for plain data types, so this should work in a multi-threaded setting. However ... if one thread is in the middle of the `transcoder_init()` (so `s_initialized` not `true` yet) and another thread calls it as well, it will result in a data race. Which leads me to think this should really be in `initialize()` instead, as that's called just once during plugin load and not in every plugin constructor (which might be called simultaenously from multiple threads).

Correct me if I'm wrong of course :)

Reply:
Sounds good to me, moved it to `initialize()` 👍